feat(linter): configure by category in config files

DonIsaac opened the pull request:
> closes #5454

Adds a `categories` property to config files, where each key is a `RuleCategory` and each value is `"allow"/"off"`, `"warn"`, or `"deny"/"error"`.

Note that this change won't come into effect until after #6088 is merged.

DonIsaac commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6120?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#6120** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6120?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#6119** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6119?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @DonIsaac and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

Boshen reviewed (commented):
To be honest I haven't seen a single usage of these categories in the wild ... since it's a foreign concept from Rust.

Should we actually try and reduce complexity at this stage?

Or maybe it's just that there is no recommended usage documentation for these categories.

DonIsaac commented:
> To be honest I haven't seen a single usage of these categories in the wild ... since it's a foreign concept from Rust.
1. [5 our of 8 non-skipped repos we test oxlint-ecosystem-ci use these categories.](https://github.com/oxc-project/oxlint-ecosystem-ci/blob/main/matrix.json).
2. The way they are using these flags will break once #6088 merges. This PR addresses the same user story that CLI flags (at least, the way these repos and many others are using them) used to solve.
3. I've seen various issues and Discord messages asking for this feature. I think it's something people want.
4. This would allow users who rely on the VSCode extension to perform broad-category rule configuration, something that is not currently possible (you can't configure CLI flags when running the LSP).
 
> Should we actually try and reduce complexity at this stage?
Yes. I have two thoughts on this:
1. This integrates directly with existing filters. It's mostly just a change to `Oxlintrc`
2. `Oxlintrc` is meant to mimic eslintrc, so we can become backwards-compatible. In the future I see us having our own, more refined, config format.
3. That being said, we should be careful what features we add to this config file

DonIsaac committed:
feat(linter): configure by category in config files (#6120)

> closes #5454

Adds a `categories` property to config files, where each key is a `RuleCategory` and each value is `"allow"/"off"`, `"warn"`, or `"deny"/"error"`.

Note that this change won't come into effect until after #6088 is merged.
